numpy/lib/function_base.py

```diff
@@ -3792,16 +4027,135 @@ def _quantile_ureduce_func(a, q, axis=None, out=None, overwrite_input=False,
                 else:
                     r = a.dtype.type(np.nan)
             else:
-                r[..., n.squeeze(0)] = a.dtype.type(np.nan)
+                r[n.squeeze(axis)] = a.dtype.type(np.nan)
         else:
             if r.ndim == 1:
                 r[:] = a.dtype.type(np.nan)
             else:
-                r[..., n.repeat(q.size, 0)] = a.dtype.type(np.nan)
+                r[..., n.squeeze(axis)] = a.dtype.type(np.nan)
 
     return r
 
 
+def _indices_for_weighted_quantile(ap, q, axis, weights):
+    """
+    Computes the indices of weighted quantiles along axis -1.
+
+    Parameters
+    ----------
+    ap : array_like
+        Array where weighted quantiles are evaluated along axis.
+    q : array_like
+        Quantiles sought.
+    axis : int
+        Axis along which to evaluate quantiles.  Will be moved to -1 for
+        np.vectorize operations.
+    weights : array_like
+        An array of weights associated with the values in `ap`.
+
+    Returns
+    -------
+    ap : array_like
+        An expanded twice as large as the input `ap` array to perform
+        interpolation on.  The axis of interest is moved to -1.
+    indices : array_like
+        Indices along the -1 axis of ap to interpolate.  All floats.
+    Nx : scalar
+        Length of the axis of ap.
+    """
+    # first move to axis -1 for np.vectorize
+    ap = np.swapaxes(ap.astype('f8'), axis, -1)
+    weights = np.swapaxes(weights.astype('f8'), axis, -1)
+
+    # values with weight=0 are made nan and later sorted to end of array
+    if ap[weights == 0.].any():
```

Comment:
You realize this means `(ap[weights == 0] != 0).any()`, right? Is this deliberate?

Reply:
Good catch.  The goal is just to find the places where weights==0, and set the corresponding cells in ap to nan.